test/remote/totp_tests.js

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const assert = require('insist')
+const config = require('../../config').getProperties()
+const TestServer = require('../test_server')
+const Client = require('../client')()
+const QrCode = require('qrcode-reader')
+const Jimp = require('jimp')
+const P = require('bluebird')
+const url = require('url')
+const querystring = require('querystring')
+const otplib = require('otplib')
+
+P.promisify(Jimp.read)
+
+function getSharedSecret(qrCodeUrl) {
+  // This takes the DataURL and decodes it to the sharedSecret
+  // used to calculate the TOTP token.
+  const qr = new QrCode()
+  const buffer = Buffer.from(qrCodeUrl.replace(/^data:image\/png;base64,/, ''), 'base64')
+
+  return Jimp.read(buffer)
+    .then((image) => {
+      return new P((resolve, reject) => {
+        qr.callback = function (err, value) {
+          if (err) {
+            reject(err)
+          }
+          resolve(value.result)
+        }
+        qr.decode(image.bitmap)
+      })
+    })
+    .then((data) => {
+      const parsedUrl = url.parse(data)
+      return querystring.parse(parsedUrl.query).secret
+    })
+}
+
+describe('remote totp', function () {
+  let server, client, email, sharedSecret, totpToken
+  const password = 'pssssst'
+
+  this.timeout(10000)
+
+  before(() => {
+    return TestServer.start(config)
+      .then(s => {
+        server = s
+      })
+  })
+
+  beforeEach(() => {
+    email = server.uniqueEmail()
+    return Client.createAndVerify(config.publicUrl, email, password, server.mailbox)
+      .then((x) => {
+        client = x
+        assert.ok(client.authAt, 'authAt was set')
+        return client.createTotpToken()
+          .then((result) => totpToken = result)
+      })
+  })
+
+  it('should create totp token', () => {
+    assert.ok(totpToken)
+    assert.ok(totpToken.qrCodeUrl)
+  })
+
+  it('should fail to create second totp token for same user', () => {
+    return client.createTotpToken()
+      .then(assert.fail, (err) => {
+        assert.equal(err.code, 400, 'correct error code')
+        assert.equal(err.errno, 154, 'correct error errno')
+      })
+  })
+
+  it('should delete totp', () => {
+    return client.deleteTotpToken()
+      .then((result) => {
+        assert.ok(result, 'delete totp token successfull')
+
+        // Can create a new token
+        return client.createTotpToken()
+          .then((result) => {
+            assert.ok(result)
+          })
+      })
+  })
+
+  it('should fail to verify totp code', () => {
+    return client.checkTotpCode('wroonn')
+      .then((result) => {
+        assert.equal(result.success, false, 'failed')
+      })
+  })
+
+  it('should verify totp code', () => {
```

Comment:
This test is flaky, which I believe has to do with failing to correctly decode the shared secret from QR code image.

Reply:
Fixed the flakiness of this test by having a configurable shared secret which removes the need to decode the return QR code.